Add 4.6 Monitoring Release Notes (#26223)
openshift/openshift-docs: release_notes/ocp-4-6-release-notes.adoc

removes the Report from your cluster at the end of its retention period. For more information,
see metering Reports xref:../metering/reports/metering-about-reports.adoc#metering-expiration_metering-about-reports[expiration].

[id="ocp-4-6-monitoring"]
=== Monitoring

[id="ocp-4-6-monitoring-for-user-defined-projects"]
==== Monitoring for user-defined projects

> In {product-title} 4.6, you can enable monitoring for user-defined projects in addition to the default platform monitoring. This facilitates monitoring for your own projects in {product-title} without the need for an additional monitoring solution. This also centralizes monitoring for core platform components and user-defined projects.

Author: @kalexand-rh thank you for your review. I have now resolved this suggestion in my local clone and will push the change to my fork shortly.

> In {product-title} 4.6, you can enable monitoring for user-defined projects in addition to the default platform monitoring. You can now monitor your own projects in {product-title} without the need for an additional monitoring solution. Using this new feature centralizes monitoring for core platform components and user-defined projects.

Reviewer: Using this new feature centralizes monitoring for **both** core platform components and user-defined projects.

Add both can make it more clear

Author: Thank you for your suggestion. My view is that the word 'both' is superfluous in this sentence.

> With this new feature, you can do the following:

Author: Thank you for highlighting that. I will keep that in mind in the future. Resolved.

* Enable and configure monitoring for user-defined projects

> * Create recording and alerting rules that use metrics from your own Pods and services

Author: Great! I have updated this.

> * Access metrics and alert information through a single, multi-tenant interface

Reviewer: Access metrics and alerts information

Supose alert here should be alerts

Author: Thank you for your suggestion. I have subsequently changed that bullet point as follows:

* Access metrics and information about alerts through a single, multi-tenant interface

* Cross-correlate the metrics for user-defined projects with platform metrics

// The following xref (../monitoring/understanding-the-monitoring-stack.adoc#understanding-the-monitoring-stack[Understanding the monitoring stack]) is dependent on https://github.com/openshift/openshift-docs/pull/24722 being merged. This will be added after the merge:

> For more information, see Understanding the monitoring stack

Reviewer 2: should link to the doc for `Understanding the monitoring stack`

[id="ocp-4-6-monitoring-prometheus-rule-validation"]
==== Prometheus rule validation

{product-title} 4.6 introduces validation of Prometheus rules through a webhook that calls the validating admission plug-in. With this enhancement, PrometheusRule custom resources in all projects are checked against the Prometheus Operator rule validation API.

[id="ocp-4-6-monitoring-metrics-and-alerting-rules-added-for-thanos-querier"]
==== Metrics and alerting rules added for the Thanos Querier

> The Thanos Querier aggregates and optionally deduplicates core OpenShift Container Platform metrics and metrics for user-defined projects under a single, multi-tenant interface. In {product-title} 4.6, a service monitor and alerting rules are now deployed for the Thanos Querier. This enables monitoring of the Thanos Querier by the monitoring stack.

Author: Resolved.

+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1874754[*BZ#1874754*])

> * In {product-title} 4.6, the resource metrics API server provides support for custom metrics. The resource metrics API server does not implement the OpenAPI specification, and the following messages are sent to the `kube-apiserver` logs every 60 seconds:

Reviewer 2: from https://bugzilla.redhat.com/show_bug.cgi?id=1819053#c1
the message frequency is every 120 seconds

Author: Thank you for highlighting this. I have now removed 'every 60 seconds' as advised.

+
[source,terminal]
----
controller.go:114] loading OpenAPI spec for "v1beta1.metrics.k8s.io" failed with: OpenAPI spec does not exist
controller.go:127] OpenAPI AggregationController: action for item v1beta1.metrics.k8s.io: Rate Limited Requeue.
----
+
In some cases, these errors might cause the `KubeAPIErrorsHigh` alert to fire, but the underlying issue is not known to degrade {product-title} functionality.

> (link:https://bugzilla.redhat.com/show_bug.cgi?id=1819053[*BZ#1819053*])

Reviewer 2: see from https://github.com/openshift/openshift-docs/blob/10eedea2c98357298b174f297b43400c17c9dffb/release_notes/ocp-4-6-release-notes.adoc
for example, (BZ#1821771) is in separate line, do we need to do the same for (BZ#1819053) and (BZ#1870287)?

Author: Thank you for highlighting this. Looking at the 'Known issues' section in the Release Notes for the previous release, each of the `link` entries are included directly after the description, without a space.

Looking at the preview, this renders the BZ link at the end of the same line as the text. Looking at [the published 4.5 Release Notes Known Issues section](https://docs.openshift.com/container-platform/4.5/release_notes/ocp-4-5-release-notes.html#ocp-4-5-known-issues), this seems to be the convention. I will leave it as is based on that.

+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1874754[*BZ#1874754*])

> * In {product-title} 4.6, the custom metrics API server provides support for custom metrics. The custom metrics API server does not implement the OpenAPI specification and the following messages are sent to the `kube-apiserver` logs every 60 seconds:

Reviewer 3: replace `custom metrics API server` with `resource metrics API server`

Author: Thank you Sergiusz for your review! I have now implemented this suggestion.

> * In {product-title} 4.6, the resource metrics API server provides support for custom metrics. The resource metrics API server does not implement the OpenAPI specification and the following messages are sent to the `kube-apiserver` logs every 60 seconds:

Author: Resolved.

+
[source,terminal]
----
controller.go:114] loading OpenAPI spec for "v1beta1.metrics.k8s.io" failed with: OpenAPI spec does not exist
controller.go:127] OpenAPI AggregationController: action for item v1beta1.metrics.k8s.io: Rate Limited Requeue.
----
+

> In some cases, the above errors might cause the `KubeAPIErrorsHigh` alert to fire.

Reviewer 4: ```suggestion
In some cases, these errors might cause the `KubeAPIErrorsHigh` alert to fire, but the underlying issue does not degrade {product-title} functionality.
```

> 🚫  above adj
Do not use to indicate a relative location in a document, as in “the above restrictions.” Use “previous” or “preceding.”

Author: Thank you. I have now combined the two sentences. I have reworded the second one slightly, because I think it is important to not to state absolutely that this does not degrade functionality:

"In some cases, these errors might cause the `KubeAPIErrorsHigh` alert to fire, but the underlying issue is not known to degrade {product-title} functionality."

+

> There is no known degradation of {product-title} functionality caused by this issue.

Reviewer 4: > ... avoid using expletive constructions such as it is, there are, and there is, which hide the subject of the sentence.

I'd combine these sentences.

Author: Resolved.

(link:https://bugzilla.redhat.com/show_bug.cgi?id=1819053[*BZ#1819053*])

> * Rules API back-ends are sometimes missed if Store API stores are detected before Rules API stores. When this occurs, a store reference is created without a Rules API client and the Rules API endpoint from Thanos Querier does not return any rules.

Reviewer 4: ```suggestion
* Rules API back-ends are sometimes missed if Store API stores are detected before Rules API stores. When this occurs, a store reference is created without a Rules API client, and the Rules API endpoint from Thanos Querier does not return any rules.
``` 

Is there a better term than "missed?"

Author: @kalexand-rh I have changed this to the following:

"Rules API back-ends are sometimes not detected if Store API stores are discovered before Rules API stores."

[id="ocp-4-6-monitoring"]
=== Monitoring

[id="ocp-4-6-monitoring-for-user-defined-projects"]
==== Monitoring for user-defined projects

In {product-title} 4.6, you can enable monitoring for user-defined projects in addition to the default platform monitoring. You can now monitor your own projects in {product-title} without the need for an additional monitoring solution. Using this new feature centralizes monitoring for core platform components and user-defined projects.

With this new feature, you can perform the following tasks:

* Enable and configure monitoring for user-defined projects
* Create recording and alerting rules that use metrics from your own pods and services
* Access metrics and information about alerts through a single, multi-tenant interface
* Cross-correlate the metrics for user-defined projects with platform metrics

For more information, see xref:../monitoring/understanding-the-monitoring-stack.adoc#understanding-the-monitoring-stack[Understanding the monitoring stack].

[id="ocp-4-6-monitoring-alerting-rule-changes"]
==== Alerting rule changes

> The following alerting rule changes have been introduced in {product-title} 4.6:

Reviewer 4: ```suggestion
The following alerting rule changes are introduced in {product-title} 4.6:
```
This is true for every instance of "has been" in this section - use the simplest tense you can.

Author: Thank you for highlighting this. Resolved.

* The `PrometheusOperatorListErrors` alert has been added. The alert provides notification of errors when running list operations on controllers.
* The `PrometheusOperatorWatchErrors` alert has been added. The alert provides notification of errors when running watch operations on controllers.

> * The `KubeQuotaExceeded` alert has changed to `KubeQuotaFullyUsed`. Previously, the `KubeQuotaExceeded` alert fired if a ResourceQuota exceeded a 90% threshold. The `KubeQuotaFullyUsed` alert fires if a ResourceQuota is fully used.

Reviewer 4: I think your use of K8 terms in this section goes against the new guidance on avoiding camel case when talking about general objects.

Author: Thank you. I have now resolved this.

* etcd alerts now support the addition of custom labels for metrics.

> * The `KubeAPILatencyHigh` and `KubeAPIErrorsHigh` alerts have been replaced by the `KubeAPIErrorBudgetBurn` alert. `KubeAPIErrorBudgetBurn` combines API error and latency alerts and only fires when the conditions are severe enough.

Reviewer 4: ```suggestion
* The `KubeAPILatencyHigh` and `KubeAPIErrorsHigh` alerts have been replaced by the `KubeAPIErrorBudgetBurn` alert. `KubeAPIErrorBudgetBurn` combines API error and latency alerts and fires only when the conditions are severe enough.
```

Author: Resolved.

> * The readiness and liveness probe metrics exposed by the kubelet are now scraped.

Reviewer 4: For what?

Author: This provides historical liveness and readiness data for containers, which can be helpful when troubleshooting container issues. I have now added this to the bullet point. Thanks for highlighting this!

> * The alerting rules for the Thanos Ruler have been updated so that alerts are paged if recording rules and alerting rules are not properly evaluated. This ensures that critical alerts are not lost when rule and alert evaluation in the Thanos Ruler is not completed.

Reviewer 4: ```suggestion
* The alerting rules for the Thanos Ruler have been updated so that alerts are paged if recording rules and alerting rules are not correctly evaluated. This update ensures that critical alerts are not lost when rule and alert evaluation in the Thanos Ruler is not completed.
```
> ! proper adj
Do not use to mean “correct,” “applicable,” or “appropriate.”

Author: Resolved.
